File: Code/12_GPIO_Polling/Ex5.c
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/poll.h>
#include <unistd.h>

#define N 100
#define Nlim (N*3)/4

void poll_btn(struct pollfd *pfd)
{
	char buffer, vals[N] = {0};
	int pos = 0, test=1, i, s;
	lseek(pfd->fd, 0, SEEK_SET);
	read(pfd->fd, &buffer, 1);
	for(i=0;i<N;i++) vals[i] = buffer-'0';
	poll(pfd, 1, -1);
	while(test)
	{
		lseek(pfd->fd, 0, SEEK_SET);
		read(pfd->fd, vals+pos, 1);
		vals[pos] -= '0';
		pos = (pos+1)%N;
		for(s=i=0; i<N; i++)
			s += vals[i];
		if(buffer=='0')
			test = s<Nlim;
		else
			test = s>Nlim;
	}
}

int main(void)
{
	struct pollfd pfd;
	int btn_press;
	system("echo 4     > /sys/class/gpio/export");
	system("echo both  > /sys/class/gpio/gpio4/edge");
	system("echo in    > /sys/class/gpio/gpio4/direction");
	pfd.fd = open("/sys/class/gpio/gpio4/value", O_RDONLY);
	if(pfd.fd < 0)
	{
		puts("Erro abrindo /sys/class/gpio/gpio4/value");
		puts("Execute este programa como root.");
		return 1;
	}
	pfd.events = POLLPRI | POLLERR;
	pfd.revents = 0;
	puts("Pressione o botao...");
	for(btn_press=0; btn_press<5; btn_press++)
	{
		poll_btn(&pfd);
		puts("Borda de descida!");
		poll_btn(&pfd);
		puts("Borda de subida!");
	}
	close(pfd.fd);
	system("echo 4 > /sys/class/gpio/unexport");
	return 0;
}

File: 3_Programação_RPi/3.3_GPIO_Polling/Python/Ex5.py
import RPi.GPIO as gpio
import os, select, time

def poll_btn(pfd, fd):
	N = 50
	Nlim = N/2
	os.lseek(fd, 0, os.SEEK_SET)
	buf = int(os.read(fd, 1))
	vals = [buf]*N
	pfd.poll()
	test=True
	pos = 0
	while test:
		os.lseek(fd, 0, os.SEEK_SET)
		vals[pos] = int(os.read(fd, 1))
		pos = (pos+1) % N
		s = sum(vals)
		if buf==0:
			test = s<Nlim
		else:
			test = s>Nlim


pin = 4
os.system("echo %d      > /sys/class/gpio/export" % pin)
os.system("echo falling > /sys/class/gpio/gpio%d/edge" % pin)
os.system("echo in      > /sys/class/gpio/gpio%d/direction" % pin)

fd = os.open("/sys/class/gpio/gpio%d/value" % pin, os.O_RDONLY)
pfd = select.poll()
pfd.register(fd, select.POLLPRI | select.POLLERR)

print("Pressione o botao...")
for btn_press in range(5):
	poll_btn(pfd, fd)
	print("Borda de descida!")
	poll_btn(pfd, fd)
	print("Borda de subida!")

os.close(fd)
os.system("echo %d > /sys/class/gpio/unexport" % pin)
